content/blog/2019-12-18-ml-based-phases-ordering-with-llvm/index.md

```diff
@@ -0,0 +1,75 @@
+
+# Design Overview
+
+## Analysis
+
+We first analyzed how much the ordering of optimization passes affects program performances. We selected the 270 optimizations used in LLVM `-O2` level as the set of optimizations we analyzed across the whole project. We performed all optimizations on the streamcluster program in PARSEC benchmark.
+
+![Performances using randomly selected passes](./w.png)
+
+For each n optimization passes n, we uniformly randomly select n optimization passes with replacement (so there can be duplicate paths) and run 10 times to obtain a range of performance. From the figure, we see that the ordering of optimizations does affect the overall program performance significantly. In fact, some order of optimizations even makes it significantly worse. As the number of passes used increases, we can see that the optimal performance improvement indeed increases but saturated around 100 in terms of the best optimization performances. Thus we can find a smaller and better optimization sequence.
+
+## Methods
+
+### Hill Climbing
+
+One popular approach for phase ordering is Hill Climbing: selecting the optimization that gives the most performance boost at each step. However, it can be really expensive to rebuild and run 270 times at every step, while also rerun the program multiple times to reduce performance variance. In fact, it could take more than a day to build even 15 passes. In the experiment, we sampled 5 optimizations at each step instead of trying all and only used one run performance to save the time.
+
+### Machine Learning Model
+
+We built a simple linear regression model to predict how much one optimization pass will improve over the performance of the current program, based on a set of performance counters we selected from ones provided by [likwid-perfctr](https://github.com/RRZE-HPC/likwid). During inference time, we will run the model to approximate Hill Climbing algorithm, i.e. greedily select the best pass based on model prediction at each step, until we reach the maximum number of passes limit or there are no more optimizations that will improve the performance.
+
+To collect data points of program performance counters, optimization pass and performace difference, we randomly sampled optimization sequences as in the analysis but with random length. We then collect the runtimes and performance counter values without the last optimization as features. The performance counters we used are:
+
+Clock per Instruction, Branch Misprediction Ratio, Instructions per Branch, % Cycles without Execution, % Cycles Without Execution due to L1D, % Cycles Without Execution due to L2, % Cycles Without Execution due to memory loads, Packed Single Precision MFLOP/S, Packed Double Precision MFLOP/S
+
+We then fit a linear regression model to predict normalized performance improvement. 
```

Comment:
OK, from this it sounds like the regression is indeed using performance counters to predict the performance. But if you have already run the benchmark, can't you just use the running time directly?

Or perhaps the setup here is that you pick an optimization sequence S, measure the performance using S, then measure, for all passes P, the performance of S+P. Then you construct a feature vector consisting of the performance counters when running using S *and the identifier of P*. That lets you predict the performance of S+P while only using an execution of S itself.

That sounds reasonable, if it's what you're doing! But I think that could be clearer here.

Reply:
Yes our setting is indeed picking a sequence S+ P (one  more opt pass), measure the performance counters from running S as the feature vector, then predict the performance difference from S to S+ P. The name of P is also concatenated to the feature vector as a categorical feature.